util/purl/image.go

```diff
@@ -86,6 +86,21 @@ func PURLToRef(purl string) (string, *ocispecs.Platform, error) {
 			if err != nil {
 				return "", nil, err
 			}
+
+			// OS-version and OS-features are not included when serializing a
```

Comment:
The problem in here looks pretty clearly a bug in the new version of `platforms.Parse()` where it does not do the job of parsing the string to a struct https://github.com/containerd/containerd/blame/main/platforms/platforms.go#L240-L253 . We shouldn't try to patch this by covering up these side-effects in the caller to make the test pass. There are 10+ calls of `platforms.Parse()` in buildkit. The only reason this showed up in purl test is likely that the integration tests are not running on windows atm. (might be missing some coverage as well).

I think the containerd patches should be reverted and if a string can be parsed to a struct with `OSVersion/OSFeatures` then the first step is to define how `OSVersion` is represented in string form.

cc @dmcgowan

Reply:
Which PR specifically? Seems this should be easy to unit test but for some reason none of these have unit tests. Microsoft folks last changed this code, we can request better test coverage around it. I don't know enough about what the behavior should even be for these fields.

Reply:
Changes to `parser.Parse` in https://github.com/containerd/containerd/pull/8778 . They read the configuration of the current system instead of parsing the input, so the return value will always be different depending on the caller's operating system or windows versions.

Reply:
Yes, I agree that the latest changes to `Parse()` are far from ideal. It's digging in deeper into an already taken direction, where `Parse()` doesn't _just parse_ but also "fills in the missing pieces" by applying defaults that are based on the host;

- `amd64` (arch only) gets "parsed" into `linux/amd64`, `darwin/amd64`, `windows/amd64` (etc) depending on the host
- `linux` (os only) gets "parsed" into `linux/amd64`, `linux/arm64` (etc), depending on the host
- and now `windows` gets "parsed" into `windows/<host-arch>/<host OS version>`

The first two are already somewhat problematic (depending on "what" the result is used for), but the `os-version` is (imo) even more problematic because there is no string-representation of that field.

I guess in an _ideal_ world, `Parse` would do _just that_; parse the string representation into a struct (_possibly_ with some amount of normalizing), but leave it to the consumer to fill in defaults (if any). Some options could be to;

- make `Parse` accept functional argument(s) to apply defaults (and/or normalizing)
- make `Parse` accept a struct with defaults

And _ideally_ also have a string representation of the additional fields (`os-version`, `os-features` ?), so that information can be preserved when serializing to a string.

Reply:
> The first two are already somewhat problematic (depending on "what" the result is used for), but the os-version is (imo) even more problematic because there is no string-representation of that field.

Yes, the first two could maybe be better with a more explicit API, but these are not blocking for us. These are the "incomplete" values that make it more obvious that the behavior might be undefined and if we wanted to we could just not allow these (we never document/promote them anyway and users should never use them imho). But for the new windows one, these are the full representations and there is no way to provide input where the parser would just parse the input in a stable form.

Reply:
CC @kiashok @profnandaa for visibility on this from the MSFT side.

To summarise, `Parse()` should just parse and return a struct with fields accurately populated based on the string we've just parsed. We absolutely need `OSVersion` on Windows to satisfy the matcher and fetch an image that can actually run on the host we're running on, but that value should be explicitly set (if needed)  before sending it to the matcher. We should not add defaults that can vary based on the environment we're running on, when parsing.

Reply:
> To summarise, Parse() should just parse and return a struct with fields accurately populated based on the string we've just parsed.

Correct; parsing should just parse what's in the string. The current "fill in host OS or host Architecture if missing" is something we should look at deprecating (but not currently the _highest_ priority as it's not a commonly used or advertised feature).

> We absolutely need OSVersion on Windows to satisfy the matcher and fetch an image that can actually run on the host we're running on, but that value should be explicitly set (if needed) before sending it to the matcher.

Yes, matching should be a separate concern from Parsing; in an _ideal_ world, the platform's string representation would also allow for `os_version` to be included, in which case matching should probably skip heuristics (and just do what's requested) but that's not the case yet.

But I guess a specific `MatchComparer` could be used to either fill in the missing `os_version` based on the host/environment, or to set the `os_version` before matching.

Reply:
It definitely needs to be up to the `MatchComparer` that decides how the host's details are intermingled, because the same `Platform` object can be a match or a non-match depending on the type of container to be run: Process-Isolated, HyperV-isolated, and Host-Process containers all have different rules for `OSVersion` values are selectable (They're approximately super-sets in that order, as it happens; empty `OSVersion` is semantically meaningful: "there's no OS present in this image" which only Host-Process mode can execute) and `MatchComparer` is the thing we want to vary in containerd to achieve this; I don't recall if the actual mechanism for this variation has been implemented yet though, I recall that it turns out to be a frontend (containerd/CRI, buildkit) concern and there was some issues found in a containerd/cri prototype, but I haven't looked for a while.

So the `platforms.Parse` (and presumably the reverse?) is intermingling concerns that should be in the Windows-specific `MatchComparer` implementations. (Which is already a thing: you can only currently get one via the Windows-build DefaultComparer call, as I recall; that could be nicer... And it implements the Process-Isolation logic)

It's a fair way down the track, but at some point we may also want, e.g., `--platform windows/amd64//10.0.0.19242` to be able to be used with a HyperV runtime to build Windows-ABI-incompatible images from the same Dockerfile on a single host, rather than needing to maintain a build-farm of down-level hosts. Presumably that'll rely on being able to parse an OSVersion out of the string, either in buildkit or internally to containerd.

Outside MatchComparers, purl's use of the same function has different purpose, highlighting the problem. Here, we need to specify the string form of a _specific image_ rather than an execution platform, and cannot assume the host's details; so for this code to be correct, we _need_ `OSVersion` in the canonical string format, or any purl formed for for a Windows image index reference will only be able to round-trip back to a HostProcess image, breaking the current use-case of attestations/provenance on Windows. (That said, GitHub Search says the only caller of `PurlToRef` is the BuildKit test suite, so incorrect parsing is probably a less-pressing issue than ambiguous stringification)

I assume from those current use-cases that a partial Platform value (filled in by host-provided-defaults) isn't desirable here.

Poking around at the purl docs (I haven't looked at them before AFAIK) it seems like `platform` is not currently specified for the `docker` type, but I assume we're treating it as "containerd's `platforms.Parse` implementation-defined string serialisation". I see `arch` is specified for the `oci` type, but buildkit doesn't support parsing the `oci` type, and the other `Platform` fields are not present there.

Reply:
> It's a fair way down the track, but at some point we may also want, e.g., --platform windows/amd64//10.0.0.19242 to

Yes, I don't see any other way around it. I would just make it `windows/amd64/10.0.0.19242`, `windows/arm64/10.0.0.19242`, `windows/arm64/v9/10.0.0.19242` as we know it is only used for windows. I'd also suggest that if platform is just `windows/amd64` then the `OSVersion` is not taken from host (even for matcher) but there is a defined value that is normalized to empty string. Same as `v8` for `arm64` and `v7` for `arm`.

> but I assume we're treating it as "containerd's platforms.Parse implementation-defined string serialisation

Yes, we don't define any other serialization in buildkit.